File: web/web_js/Tag.js

var SimpleTag = require('SimpleTag');

function Tag(x, y, imgSize, imgScale){
	super(x, y); // TODO change

	this.active = true;
	
	this.resizing = false;
	this.moving = false;
	
	this.text = "";
	this.textLayout  = new GlyphLayout(); // TODO change
	
	this.imgSize = imgSize;
	this.imgScale = imgScale;
	
	//where the image ends on the screen
	this.leftBoundary = (854 / 2) - (imgSize.x / 2);
	this.rightBoundary = (854 / 2) + (imgSize.x / 2);
	this.bottomBoundary = 50;

}

// TODO make Tag inherit SimpleTag and see how to do the compareTo
Tag.inherits(SimpleTag);
Tag.constructor = SimpleTag()
Tag.prototype.method = function ('toString', func) {
    this.prototype['toString'] = func;
    return this;
}

Tag.prototype.compareTo = function(obj){
	if(obj instanceof Tag){
		return  parseInt(getArea() - obj.getArea());
	}
	else{
		//TODO log 
		//Gdx.app.log("debug", obj.getClass().toString());
		return 0;
	}
	
}


Tag.prototype.toString = function(){
	return position.toString() + " " + size.toString() + " " + getArea();
}

Tag.prototype.update = function(point){
	if(this.active){
		if(this.resizing){
			if(point[0] < this.leftBoundary){
				point[0] = this.leftBoundary;
			}
			if(point[0] > this.rightBoundary){
				point[0] = this.rightBoundary;
			}
			
			if(point[1] < this.bottomBoundary){
				point[1] = 50;
			}
			if(point[1] > 480){
				point[1] = 480;
			}
			
			var tmp = point - this.position; // TODO check if this.position is accessible (super class attribute)
			
			if(tmp[0] >= 0 && tmp[0] <= 50){
				tmp[0] = 50;
			}
			if(tmp[0] < 0 && tmp[0] >= -50){
				tmp[0] = -50;
			}
			
			if(tmp[1] >= 0 && tmp[1] <= 50){
				tmp[1] = 50;
			}
			if(tmp[1] < 0 && tmp[1] >= -50){
				tmp[1] = -50;
			}
			
			this.size = tmp;
		}

		else if(this.moving){
			
			var tmp = point;
			
			if(tmp[0] < (854 / 2) - (this.imgSize[0] / 2) + Math.abs(Math.min(this.size[0], 0))){
				tmp[0] = (854 / 2) - (this.imgSize[0] / 2) + Math.abs(Math.min(this.size[0], 0));
			}
			if(tmp[0] > (854 / 2) + (this.imgSize[0] / 2) - Math.abs(Math.max(this.size[0], 0))){
				tmp[0] = (854 / 2) + (this.imgSize[0] / 2) - Math.abs(Math.max(this.size[0], 0));
			}
			
			if(tmp[1] < 50 + Math.abs(Math.min(this.size[1], 0))){
				tmp[1] = 50 + Math.abs(Math.min(this.size[1], 0));
			}
			if(tmp[1] > 480 - Math.abs(Math.max(this.size[1], 0))){
				tmp[1] = 480 - Math.abs(Math.max(this.size[1], 0));
			}
			
			this.position = tmp;
		}
	}
}

Tag.prototype.render = function(shapeRenderer){
	// TODO check how to do it
}

Tag.prototype.renderText = function(batch, bitmapFont){
	// TODO check how to do it
}

Tag.prototype.contains = function(point){
	var x, y, w, h;
	
	if(size[0] < 0){
		x = position[0] + size[0];
	}
	else{
		x = position[0];
	}
	
	if(size[1] < 0){
		y = position[1] + size[1];
	}
	else{
		y = position[1];
	}
	
	w = Math.abs(size[0]);
	h = Math.abs(size[1]);
	
	return (point[0] >= x) && (point[0] <= x + w) &&
			(point[1] >= y) && (point[1] <= y + h);
}

Tag.prototype.grabHandles = function(point){
	var tmp = this.position;

	var a1 = (point[0]-tmp[0]);
	var b1 = (point[1]-tmp[1]);
	var dst_Point_Position = Math.sqrt(a1*a1 + b1*b1);
	//var dst_Point_Position = dst(point, tmp)

	var a2 = (this.position[0]+this.size[0])-point[0];
	var b2 = (this.position[1]+this.size[1])-point[1];
	var dst_Point_PositionSize = Math.sqrt(a2*a2 + b2*b2);
	//var dst_Point_PositionSize = dst(this.position+this.size, point)

	if(dst_Point_PositionSize < 25 && !this.moving){
		this.resizing = true;
	}
	
	if(dst_Point_Position < 25 && !this.resizing){
		this.moving = true;
	}
}

function dst(point1, point2){
	var a1 = (point1[0]-point2[0]);
	var b1 = (point1[1]-point2[1]);
	return = Math.sqrt(a1*a1 + b1*b1);
}

Tag.prototype.handleGrabbed = function (){
	return this.moving || this.resizing;
}

Tag.prototype.releaseHandles = function (){
	this.moving = false;
	this.resizing = false;
}

Tag.prototype.setActive = function (flag){
	this.active = flag;
}

Tag.prototype.isActive = function (){
	return this.active;
}

Tag.prototype.getArea = function (){
	return this.size.x * this.size.y;
}

Tag.prototype.setSharkId = function (id, text){
	super.setSharkId(id);
	this.text = text;
}

Tag.prototype.getText = function (){
	return this.text;
}

Tag.prototype.toSimpleTag = function(){
	var t = new SimpleTag(0, 0);
	t.position = getOriginalPosition(this.position);
	t.size = getOriginalSize(this.size);
	t.sharkId = this.sharkId;
	
	// Convert so that position is in the top-left corner of the tag
	// and size is always positive going bottom-right
	if(t.size[1] > 0){
		t.position[1] += t.size[1];
	}
	else{
		t.size[1] *= -1;	// change to positive
	}
	
	if(t.size[0] < 0){
		t.position[0] += t.size[0];
		t.size[0] *= -1;
	}
	t.position[1] = imgSize[1] * imgScale - t.position[1];	// invert Y axis
	return t;
}

Tag.prototype.getOriginalPosition = function(point){
	var tmp = point;
	tmp[0] = (tmp[0] - leftBoundary) * imgScale;
	tmp[1] = (tmp[1] - bottomBoundary) * imgScale;
	return tmp;
}

Tag.prototype.getOriginalSize = function(point){
	var tmp[size]
	tmp[0] = tmp[0] * imgScale;
	tmp[1] = tmp[1] * imgScale;
	
	return tmp;
}

module.exports = Tag;

File: Shark-Tagging-Game/core/src/com/dcu/sharktag/Tag.java
package com.dcu.sharktag;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.GlyphLayout;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.math.Vector2;

public class Tag extends SimpleTag implements Comparable{
	
	private boolean active;
	
	private boolean resizing = false;
	private boolean moving = false;
	
	private String text = "";
	private GlyphLayout textLayout;
	
	private Vector2 imgSize;
	private float imgScale;
	
	//where the image ends on the screen
	private float leftBoundary = 0;
	private float rightBoundary = 0;
	private float bottomBoundary = 0;
	
	public Tag(float x, float y, Vector2 imgSize, float imgScale){
		
		super(x, y);

		active = true;
		textLayout = new GlyphLayout();
		this.imgSize = imgSize;
		this.imgScale = imgScale;
		
		leftBoundary = (854 / 2) - (imgSize.x / 2);
		rightBoundary = (854 / 2) + (imgSize.x / 2);
		bottomBoundary = 50;
	}
	
	@Override
	public int compareTo(Object obj) {
		
		if(obj.getClass() != Tag.class){
			Gdx.app.log("debug", obj.getClass().toString());
			return 0;
		}
		
		Tag other = (Tag)obj;
		
		return (int)(getArea() - other.getArea());
	}
	
	public String toString(){
		return position.toString() + " " + size.toString() + " " + getArea();
	}

	public void update(Vector2 point){
		if(active){
			if(resizing){
				if(point.x < leftBoundary){
					point.x = leftBoundary;
				}
				if(point.x > rightBoundary){
					point.x = rightBoundary;
				}
				
				if(point.y < bottomBoundary){
					point.y = 50;
				}
				if(point.y > 480){
					point.y = 480;
				}
				
				Vector2 tmp = point.sub(position);
				
				if(tmp.x >= 0 && tmp.x <= 50){
					tmp.x = 50;
				}
				if(tmp.x < 0 && tmp.x >= -50){
					tmp.x = -50;
				}
				
				if(tmp.y >= 0 && tmp.y <= 50){
					tmp.y = 50;
				}
				if(tmp.y < 0 && tmp.y >= -50){
					tmp.y = -50;
				}
				
				size = tmp;
			}

			else if(moving){
				
				Vector2 tmp = point;
				
				if(tmp.x < (854 / 2) - (imgSize.x / 2) + Math.abs(Math.min(size.x, 0))){
					tmp.x = (854 / 2) - (imgSize.x / 2) + Math.abs(Math.min(size.x, 0));
				}
				if(tmp.x > (854 / 2) + (imgSize.x / 2) - Math.abs(Math.max(size.x, 0))){
					tmp.x = (854 / 2) + (imgSize.x / 2) - Math.abs(Math.max(size.x, 0));
				}
				
				if(tmp.y < 50 + Math.abs(Math.min(size.y, 0))){
					tmp.y = 50 + Math.abs(Math.min(size.y, 0));
				}
				if(tmp.y > 480 - Math.abs(Math.max(size.y, 0))){
					tmp.y = 480 - Math.abs(Math.max(size.y, 0));
				}
				
				position = tmp;
			}
		}
	}
	
	public void render(ShapeRenderer shapeRenderer){
		Gdx.gl.glEnable(GL20.GL_BLEND);
		Gdx.gl.glBlendFunc(GL20.GL_SRC_ALPHA, GL20.GL_ONE_MINUS_SRC_ALPHA);
		
		shapeRenderer.begin(ShapeRenderer.ShapeType.Line);
		
		if(active){
			if(sharkId > 0){
				shapeRenderer.setColor(0, 1, 0, 1);
			}
			else{
				shapeRenderer.setColor(1, 0, 0, 1);
			}
		}
		else{
			if(sharkId > 0){
				shapeRenderer.setColor(0, 1, 0, 0.6f);
			}
			else{
				shapeRenderer.setColor(1, 0, 0, 0.6f);
			}
		}
		
		shapeRenderer.rect(position.x, position.y,
				size.x, size.y);
		
		shapeRenderer.end();
		
		shapeRenderer.begin(ShapeRenderer.ShapeType.Filled);
		if(active){
			shapeRenderer.setColor(0, 0, 1, 1);	// Debug
			shapeRenderer.circle(position.x, position.y, 5);
			
			shapeRenderer.setColor(0, 1, 0, 1);
			shapeRenderer.circle(position.x + size.x, position.y + size.y, 5);
			
			shapeRenderer.setColor(1, 0, 1, 1);
			shapeRenderer.rect(position.x - 1, position.y - 7, 0, 0, 10f, 10f, 1f, 1f, 45f);
		}
		
		Gdx.gl.glDisable(GL20.GL_BLEND);
		shapeRenderer.end();
	}
	
	public void renderText(Batch batch, BitmapFont bitmapFont){
		if(sharkId > 0){
			batch.setColor(1, 1, 1, 1);
			textLayout.setText(bitmapFont, text);
			bitmapFont.draw(batch, text,
					position.x, position.y);
		}
	}
	
	public boolean contains(Vector2 point){
		
		float x, y, w, h;
		
		if(size.x < 0){
			x = position.x + size.x;
		}
		else{
			x = position.x;
		}
		
		if(size.y < 0){
			y = position.y + size.y;
		}
		else{
			y = position.y;
		}
		
		w = Math.abs(size.x);
		h = Math.abs(size.y);
		
		return (point.x >= x) && (point.x <= x + w) &&
				(point.y >= y) && (point.y <= y + h);
	}
	
	public void grabHandles(Vector2 point){
		Vector2 tmp = new Vector2(position);
		if(point.dst(tmp.add(size)) < 25 && !moving){
			resizing = true;
		}
		
		if(point.dst(position) < 25 && !resizing){
			moving = true;
		}
	}
	
	public boolean handleGrabbed(){
		return moving || resizing;
	}
	
	public void releaseHandles(){
		moving = false;
		resizing = false;
	}
	
	public void setActive(boolean flag){
		active = flag;
	}
	
	public boolean isActive(){
		return active;
	}
	
	public float getArea(){
		return size.x * size.y;
	}
	
	public void setSharkId(int id, String text){
		super.setSharkId(id);
		this.text = text;
	}
	
	public String getText(){
		return text;
	}
	
	public SimpleTag toSimpleTag(){
		SimpleTag t = new SimpleTag(0, 0);
		t.position = getOriginalPosition(this.position);
		t.size = getOriginalSize(this.size);
		t.sharkId = this.sharkId;
		
		// Convert so that position is in the top-left corner of the tag
		// and size is always positive going bottom-right
		if(t.size.y > 0){
			t.position.y += t.size.y;
		}
		else{
			t.size.y *= -1;	// change to positive
		}
		
		if(t.size.x < 0){
			t.position.x += t.size.x;
			t.size.x *= -1;
		}
		t.position.y = imgSize.y * imgScale - t.position.y;	// invert Y axis
		return t;
	}
	
	// Convert the position and size to match the original scale of the picture
	private Vector2 getOriginalPosition(Vector2 point){
		Vector2 tmp = new Vector2(point);
		tmp.x = (tmp.x - leftBoundary) * imgScale;
		tmp.y = (tmp.y - bottomBoundary) * imgScale;
		
		return tmp;
	}
	
	private Vector2 getOriginalSize(Vector2 point){
		Vector2 tmp = new Vector2(size);
		tmp.x = tmp.x * imgScale;
		tmp.y = tmp.y * imgScale;
		
		return tmp;
	}
}
